Adaptor for std::vector doesn't support types without default constructors

I took the example from the wiki about non-default-constructible types and added a vector for the same type (any other non-default-constructible type will cause the same error):
```
namespace foo {
struct no_def_con {
    no_def_con() = delete;

    int i;
    std::vector<no_def_con> v;
    MSGPACK_DEFINE(i, v);
};
}

namespace msgpack {
    MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS) {
        namespace adaptor {
        template <>
        struct as<foo::no_def_con> {
            foo::no_def_con operator()(msgpack::object const& o) const {
                if (o.type != msgpack::type::ARRAY) throw msgpack::type_error();
                if (o.via.array.size != 2) throw msgpack::type_error();
                return foo::no_def_con{o.via.array.ptr[0].as<int>(), o.via.array.ptr[1].as<std::vector<foo::no_def_con>>()};
            }
        };
        } // adaptor
    } // MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS)
} // msgpack
```
Trying to compile this results in:
```n file included from /usr/include/c++/7.2.0/memory:64:0,
                 from /usr/include/msgpack/v1/cpp_config_decl.hpp:98,
                 from /usr/include/msgpack/cpp_config_decl.hpp:13,
                 from /usr/include/msgpack/cpp_config.hpp:13,
                 from /usr/include/msgpack/v1/pack_decl.hpp:14,
                 from /usr/include/msgpack/pack_decl.hpp:13,
                 from /usr/include/msgpack/pack.hpp:13,
                 from /usr/include/msgpack/v1/object_decl.hpp:14,
                 from /usr/include/msgpack/object_decl.hpp:14,
                 from /usr/include/msgpack/object.hpp:13,
                 from /usr/include/msgpack.hpp:10,
                 from msgpack_no_def_ctor.cpp:1:
/usr/include/c++/7.2.0/bits/stl_construct.h: In instantiation of ‘void std::_Construct(_T1*, _Args&& ...) [with _T1 = foo::no_def_con; _Args = {}]’:
/usr/include/c++/7.2.0/bits/stl_uninitialized.h:527:18:   required from ‘static _ForwardIterator std::__uninitialized_default_n_1<_TrivialValueType>::__uninit_default_n(_ForwardIterator, _Size) [with _ForwardIterator = foo::no_def_con*; _Size = long unsigned int; bool _TrivialValueType = false]’
/usr/include/c++/7.2.0/bits/stl_uninitialized.h:583:20:   required from ‘_ForwardIterator std::__uninitialized_default_n(_ForwardIterator, _Size) [with _ForwardIterator = foo::no_def_con*; _Size = long unsigned int]’
/usr/include/c++/7.2.0/bits/stl_uninitialized.h:645:44:   required from ‘_ForwardIterator std::__uninitialized_default_n_a(_ForwardIterator, _Size, std::allocator<_Tp>&) [with _ForwardIterator = foo::no_def_con*; _Size = long unsigned int; _Tp = foo::no_def_con]’
/usr/include/c++/7.2.0/bits/vector.tcc:563:35:   required from ‘void std::vector<_Tp, _Alloc>::_M_default_append(std::vector<_Tp, _Alloc>::size_type) [with _Tp = foo::no_def_con; _Alloc = std::allocator<foo::no_def_con>; std::vector<_Tp, _Alloc>::size_type = long unsigned int]’
/usr/include/c++/7.2.0/bits/stl_vector.h:692:21:   [ skipping 2 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/usr/include/msgpack/v1/adaptor/adaptor_base.hpp:58:42:   required from ‘typename std::enable_if<(! std::is_array< <template-parameter-1-1> >::value), const msgpack::v2::object&>::type msgpack::v1::operator>>(const msgpack::v2::object&, T&) [with T = std::vector<foo::no_def_con>; typename std::enable_if<(! std::is_array< <template-parameter-1-1> >::value), const msgpack::v2::object&>::type = const msgpack::v2::object&]’
/usr/include/msgpack/v1/object.hpp:545:24:   required from ‘typename std::enable_if<((! std::is_array< <template-parameter-1-1> >::value) && (! std::is_pointer<_Tp>::value)), T&>::type msgpack::v1::object::convert(T&) const [with T = std::vector<foo::no_def_con>; typename std::enable_if<((! std::is_array< <template-parameter-1-1> >::value) && (! std::is_pointer<_Tp>::value)), T&>::type = std::vector<foo::no_def_con>&]’
/usr/include/msgpack/v2/object_fwd.hpp:60:56:   required from ‘T& msgpack::v2::object::convert(T&) const [with T = std::vector<foo::no_def_con>]’
/usr/include/msgpack/v1/adaptor/detail/cpp11_define_array.hpp:34:9:   required from ‘static void msgpack::v1::type::define_array_imp<Tuple, N>::unpack(const msgpack::v2::object&, Tuple&) [with Tuple = std::tuple<int&, std::vector<foo::no_def_con, std::allocator<foo::no_def_con> >&>; long unsigned int N = 2]’
/usr/include/msgpack/v1/adaptor/detail/cpp11_define_array.hpp:75:72:   required from ‘void msgpack::v1::type::define_array<Args>::msgpack_unpack(const msgpack::v2::object&) [with Args = {int, std::vector<foo::no_def_con, std::allocator<foo::no_def_con> >}]’
msgpack_no_def_ctor.cpp:10:5:   required from here
/usr/include/c++/7.2.0/bits/stl_construct.h:75:7: error: use of deleted function ‘foo::no_def_con::no_def_con()’
     { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
msgpack_no_def_ctor.cpp:6:5: note: declared here
     no_def_con() = delete;
     ^~~~~~~~~~
```
I might be doing something stupid but I think this demonstrated that the vector adaptor always calls `convert()`. I have not tried but there are possibly other adaptors that do the same.

--- a/Files.cmake
+++ b/Files.cmake
@@ -541,6 +541,7 @@ IF (MSGPACK_ENABLE_CXX)
         include/msgpack/v1/adaptor/detail/cpp03_define_map_decl.hpp
         include/msgpack/v1/adaptor/detail/cpp03_msgpack_tuple.hpp
         include/msgpack/v1/adaptor/detail/cpp03_msgpack_tuple_decl.hpp
+        include/msgpack/v1/adaptor/detail/cpp11_convert_helper.hpp
         include/msgpack/v1/adaptor/detail/cpp11_define_array.hpp
         include/msgpack/v1/adaptor/detail/cpp11_define_array_decl.hpp
         include/msgpack/v1/adaptor/detail/cpp11_define_map.hpp

--- a/include/msgpack/v1/adaptor/detail/cpp11_convert_helper.hpp
+++ b/include/msgpack/v1/adaptor/detail/cpp11_convert_helper.hpp
@@ -1,0 +1,45 @@
+//
+// MessagePack for C++ static resolution routine
+//
+// Copyright (C) 2017 KONDO Takatoshi
+//
+//    Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//    http://www.boost.org/LICENSE_1_0.txt)
+//
+#ifndef MSGPACK_V1_CPP11_CONVERT_HELPER_HPP
+#define MSGPACK_V1_CPP11_CONVERT_HELPER_HPP
+
+#include <tuple>
+
+#include <msgpack/object_fwd.hpp>
+
+namespace msgpack {
+/// @cond
+MSGPACK_API_VERSION_NAMESPACE(v1) {
+/// @endcond
+namespace type {
+
+template <typename T>
+inline typename std::enable_if<
+    has_as<T>::value
+>::type
+convert_helper(msgpack::object const& o, T& t) {
+    t = o.as<T>();
+}
+template <typename T>
+inline typename std::enable_if<
+    !has_as<T>::value
+>::type
+convert_helper(msgpack::object const& o, T& t) {
+    o.convert(t);
+}
+
+}  // namespace type
+
+/// @cond
+}  // MSGPACK_API_VERSION_NAMESPACE(v1)
+/// @endcond
+}  // namespace msgpack
+
+#endif // MSGPACK_V1_CPP11_CONVERT_HELPER_HPP

--- a/include/msgpack/v1/adaptor/detail/cpp11_define_array.hpp
+++ b/include/msgpack/v1/adaptor/detail/cpp11_define_array.hpp
@@ -21,39 +21,6 @@ MSGPACK_API_VERSION_NAMESPACE(v1) {
 /// @endcond
 namespace type {
 
-namespace detail {
-template <int N, typename... Ts>
-struct get;
-
-template <int N, typename T, typename... Ts>
-struct get<N, std::tuple<T, Ts...>>
-{
-    using type = typename get<N - 1, std::tuple<Ts...>>::type;
-};
-
-template <typename T, typename... Ts>
-struct get<0, std::tuple<T, Ts...>>
-{
-    using type = T;
-};
-
-template <typename T>
-inline typename std::enable_if<
-    has_as<T>::value
->::type
-unpack_impl(msgpack::object const& o, T& t) {
-    t = o.as<T>();
-}
-template <typename T>
-inline typename std::enable_if<
-    !has_as<T>::value
->::type
-unpack_impl(msgpack::object const& o, T& t) {
-    o.convert(t);
-}
-
-} // namespace detail
-
 template <typename Tuple, std::size_t N>
 struct define_array_imp {
     template <typename Packer>

--- a/include/msgpack/v1/adaptor/detail/cpp11_define_array.hpp
+++ b/include/msgpack/v1/adaptor/detail/cpp11_define_array.hpp
@@ -11,6 +11,7 @@
 #define MSGPACK_V1_CPP11_DEFINE_ARRAY_HPP
 
 #include "msgpack/v1/adaptor/detail/cpp11_define_array_decl.hpp"
+#include "msgpack/v1/adaptor/detail/cpp11_convert_helper.hpp"
 
 #include <tuple>
 
@@ -19,6 +20,39 @@ namespace msgpack {
 MSGPACK_API_VERSION_NAMESPACE(v1) {
 /// @endcond
 namespace type {
+
+namespace detail {
+template <int N, typename... Ts>
+struct get;
+
+template <int N, typename T, typename... Ts>
+struct get<N, std::tuple<T, Ts...>>
+{
+    using type = typename get<N - 1, std::tuple<Ts...>>::type;
+};
+
+template <typename T, typename... Ts>
+struct get<0, std::tuple<T, Ts...>>
+{
+    using type = T;
+};
+
+template <typename T>
+inline typename std::enable_if<
+    has_as<T>::value
+>::type
+unpack_impl(msgpack::object const& o, T& t) {
+    t = o.as<T>();
+}
+template <typename T>
+inline typename std::enable_if<
+    !has_as<T>::value
+>::type
+unpack_impl(msgpack::object const& o, T& t) {
+    o.convert(t);
+}
+
+} // namespace detail
 
 template <typename Tuple, std::size_t N>
 struct define_array_imp {
@@ -31,7 +65,7 @@ struct define_array_imp {
         define_array_imp<Tuple, N-1>::unpack(o, t);
         const size_t size = o.via.array.size;
         if(size <= N-1) { return; }
-        o.via.array.ptr[N-1].convert(std::get<N-1>(t));
+        convert_helper(o.via.array.ptr[N-1], std::get<N-1>(t));
     }
     static void object(msgpack::object* o, msgpack::zone& z, Tuple const& t) {
         define_array_imp<Tuple, N-1>::object(o, z, t);
@@ -48,7 +82,7 @@ struct define_array_imp<Tuple, 1> {
     static void unpack(msgpack::object const& o, Tuple& t) {
         const size_t size = o.via.array.size;
         if(size <= 0) { return; }
-        o.via.array.ptr[0].convert(std::get<0>(t));
+        convert_helper(o.via.array.ptr[0], std::get<0>(t));
     }
     static void object(msgpack::object* o, msgpack::zone& z, Tuple const& t) {
         o->via.array.ptr[0] = msgpack::object(std::get<0>(t), z);

--- a/include/msgpack/v1/adaptor/detail/cpp11_define_map.hpp
+++ b/include/msgpack/v1/adaptor/detail/cpp11_define_map.hpp
@@ -11,6 +11,7 @@
 #define MSGPACK_V1_CPP11_DEFINE_MAP_HPP
 
 #include "msgpack/v1/adaptor/detail/cpp11_define_map_decl.hpp"
+#include "msgpack/v1/adaptor/detail/cpp11_convert_helper.hpp"
 
 #include <tuple>
 #include <map>
@@ -34,7 +35,7 @@ struct define_map_imp {
         define_map_imp<Tuple, N-2>::unpack(o, t, kvmap);
         auto it = kvmap.find(std::get<N-2>(t));
         if (it != kvmap.end()) {
-            it->second->convert(std::get<N-1>(t));
+            convert_helper(*it->second, std::get<N-1>(t));
         }
     }
     static void object(msgpack::object* o, msgpack::zone& z, Tuple const& t) {

